.github/workflows/e2e-tests.yml

```
@@ -197,8 +197,8 @@ jobs:
   test-report:
     name: "Test report"
     runs-on: ubuntu-latest
-    if: ${{ ! cancelled() }}
-    needs: [create-test-matrix, e2e-tests]
+    if: ${{ needs.e2e-tests.result == 'success' || needs.e2e-tests.result == 'failure' }}
```

What can the results be, if not success or failure?

---

There is `cancelled` and I suspect another status, for when the job doesn't run at all, but is in waiting, which is the case when the matrix returns empty. 
The initial `!cancelled()` condition was overwriting the `needs` condition, making the reporting jobs run everytime the entire workflow was anything else but cancelled until that point.

---

Statuses are `success`, `failure`, `cancelled`, or `skipped` (see [here](https://docs.github.com/en/actions/learn-github-actions/contexts#needs-context)).

This will have the opposite problem: unless you use one of the [status check functions](https://docs.github.com/en/actions/learn-github-actions/expressions#status-check-functions), Actions will assume the action should only run if all the `needs` jobs succeeded. So it will only run for `success`, not `failure`.

You could try doing like this
```suggestion
    if: ${{ success() || failure() }}
```
although it might consider "cancelled" as a failure. Or you could just include a dummy condition like
```suggestion
    if: ${{ always() && ( needs.e2e-tests.result == 'success' || needs.e2e-tests.result == 'failure' ) }}
```
to signal that you don't want the default implicit `success()` check.

---

> This will have the opposite problem: unless you use one of the [status check functions](https://docs.github.com/en/actions/learn-github-actions/expressions#status-check-functions), Actions will assume the action should only run if all the needs jobs succeeded. So it will only run for success, not failure.

Oh, interesting, I didn't catch that.

From my few tests, using the `always() ...` version will also run it when the matrix is empty, and no tests ran, which is the same as the original problem.
I think using status check functions `if: ${{ success() || failure() }}` works as expected.

---

Updated with [69fc3b4](https://github.com/Automattic/jetpack/pull/28913/commits/69fc3b462433405a8954093cf701b83a493b0304)

Validated with these runs: [cancelled](https://github.com/Automattic/jetpack/actions/runs/4168417816), [failure](https://github.com/Automattic/jetpack/actions/runs/4168477803/attempts/1), [cancelled again](https://github.com/Automattic/jetpack/actions/runs/4168477803/attempts/2), [success](https://github.com/Automattic/jetpack/actions/runs/4168477803/attempts/3)